jekyll/_posts/2018-06-03-create-a-simple-macro.md

```diff
@@ -0,0 +1,120 @@
+
+---
+title: "Create a simple macro"
+author: Antonis Geralis
+---
+
+Hello, as you might know Nim is a powerful programming language that supports
+metaprogramming using macros. Though a lot of Nim programmers are unaware about
+the merits of them due to lack of learning resources. In the first part of
+this series we will discuss the use of macros to simplify the creation of
+boilerplate code in Nim.
+
+Suppose we have code which builds a directed graph.
+
+```nim
+  proc buildCityGraph(): Digraph =
+    result = initGraph()
+    result.addEdge(initEdge(initNode("Boston"), initNode("Providence")))
+    result.addEdge(initEdge(initNode("Boston"), initNode("New York")))
+    result.addEdge(initEdge(initNode("Providence"), initNode("Boston")))
+    result.addEdge(initEdge(initNode("Providence"), initNode("New York")))
+    ...
+```
+
+A template could be used to reduce the amount of typing. For example:
+
+```nim
+   template adder(graph, src, dest): untyped =
+      graph.addEdge(initEdge(initNode(src), initNode(dest)))
+```
+
+However it needs to be called each time with all three arguments. This excludes
+the possibility of using an operator with a nice syntax. Although operators can
+have more than two parameters, we are limited into calling them like
+this: `` `->`(result, "Boston", "Providence") ``. A macro, however can be used
+to create a simpler and more appealling syntax.
```

Comment:
Actually a template would work for this too. I understand that you want to show off macros here, so add maybe "(A template could do so too here, but I want to show how macros work.)"

Reply:
Just curious, you mean, result is hardcoded in the template?